### PR TITLE
CircleCI NPM Publishing

### PR DESCRIPTION
**Pull request by bucktower**
Now publishing to https://www.npmjs.com/package/@silvercar/js-constable harnessing CircleCI

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    steps:
+      - checkout
+      - attach_workspace:
+          at: ~/constable
+      - run:
+          name: Authenticate with registry
+          command: echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > ~/constable/.npmrc
+      - run:
+          name: Publish package
+          command: npm publish --access public
```
Comment (reneemeyer) [inline .circleci/config.yml]: This job should only happen on master merge. Can you add the branch specification?

**Review by reneemeyer — CHANGES_REQUESTED** [review]

**Comment by bucktower** [comment]
@reneedudley It won't let me rerequest review, but fixed to only look for master and not running anymore on this branch.

Screen Shot 2019-11-21 at 3 39 20 PM

**Review by reneemeyer — APPROVED** [review]

**Comment by johnny-1111one1** [comment]
@bucktower Why are we publishing when merging into master? Shouldn't we publish as part of a release? How is versioning being handled? If I open a PR that fixes a typo, does that overwrite the existing release? We don't need to release a typo fix on its own, either.

We need to make sure that we're purposefully iterating the library, from `1.0.0` -> `1.0.1` or `1.1.0`, etc, without impacting any existing versions of the library out there.

I can't tell if the expectation was that everyone referenced the latest version. That could mean that changing the linting file adds a rule that no clients are following, then all the clients start breaking on their next build.

**Comment by bucktower** [comment]
> @bucktower Why are we publishing when merging into master? Shouldn't we publish as part of a release? How is versioning being handled? If I open a PR that fixes a typo, does that overwrite the existing release? We don't need to release a typo fix on its own, either.
> 
> We need to make sure that we're purposefully iterating the library, from `1.0.0` -> `1.0.1` or `1.1.0`, etc, without impacting any existing versions of the library out there.
> 
> I can't tell if the expectation was that everyone referenced the latest version. That could mean that changing the linting file adds a rule that no clients are following, then all the clients start breaking on their next build.

@johnny-1111one1 I figured we'd use [semver](https://semver.org/) for versioning. I figured trunk-based branching with continuous deployment off of `master` was the quickest way to make sure we stay up-to-date with PRs for the checks. If there is an entire pull request etc off of a typo or updated documentation, I don't see what the reasoning is for not releasing it immediately or the downside of just incrementing a version esp when it isn't breaking.

However, if the majority of people want to do release branches and all that, that's fine, but in that case, I think that doing git flow with `develop` and `master` branches make more sense than merging into `master`, then branching off of master into `rc-**` and then merging `rc-**` back into `master` off of a release and potentially dealing with even more breaking changes and merge conflicts introduced into master before `rc-**` is merged back in. If we're going to use the git-flow model, continuous deployment off of master branch still makes sense as there is always one source of truth to go to for what code (and README) is live. No cross-referencing.

Perhaps I could find a way to automatically tag `package.json` w up-to-date version using git tags?

**Comment by johnny-1111one1** [comment]
> > @bucktower Why are we publishing when merging into master? Shouldn't we publish as part of a release? How is versioning being handled? If I open a PR that fixes a typo, does that overwrite the existing release? We don't need to release a typo fix on its own, either.
> > We need to make sure that we're purposefully iterating the library, from `1.0.0` -> `1.0.1` or `1.1.0`, etc, without impacting any existing versions of the library out there.
> > I can't tell if the expectation was that everyone referenced the latest version. That could mean that changing the linting file adds a rule that no clients are following, then all the clients start breaking on their next build.
> 
> @johnny-1111one1 I figured we'd use [semver](https://semver.org/) for versioning. I figured trunk-based branching with continuous deployment off of `master` was the quickest way to make sure we stay up-to-date with PRs for the checks. If there is an entire pull request etc off of a typo or updated documentation, I don't see what the reasoning is for not releasing it immediately or the downside of just incrementing a version esp when it isn't breaking.
> 
> However, if the majority of people want to do release branches and all that, that's fine, but in that case, I think that doing git flow with `develop` and `master` branches make more sense than merging into `master`, then branching off of master into `rc-**` and then merging `rc-**` back into `master` off of a release and potentially dealing with even more breaking changes and merge conflicts introduced into master before `rc-**` is merged back in. If we're going to use the git-flow model, continuous deployment off of master branch still makes sense.
> 
> Perhaps I could find a way to automatically tag `package.json` w up-to-date version using git tags?

I don't think we need release branches. I don't think merging into master should publish a new version of the library. Instead, we can have CircleCI publish to npm when we make a release. So we create 1.0.0, release that as the base version. Then we make a few changes (let's say 3), that's 3 PRs into master. We decide we want that to be 1.0.1, so we release the current master branch with release 1.0.1, which gets published to npm. Now npm has versions 1.0.0 and 1.0.1. I don't think we need to release every PR or merge into master.

This workflow is how I setup the Ruby linter: https://github.com/silvercar/silvercop/blob/master/.circleci/config.yml

**Comment by bucktower** [comment]
> > > @bucktower Why are we publishing when merging into master? Shouldn't we publish as part of a release? How is versioning being handled? If I open a PR that fixes a typo, does that overwrite the existing release? We don't need to release a typo fix on its own, either.
> > > We need to make sure that we're purposefully iterating the library, from `1.0.0` -> `1.0.1` or `1.1.0`, etc, without impacting any existing versions of the library out there.
> > > I can't tell if the expectation was that everyone referenced the latest version. That could mean that changing the linting file adds a rule that no clients are following, then all the clients start breaking on their next build.
> > 
> > 
> > @johnny-1111one1 I figured we'd use [semver](https://semver.org/) for versioning. I figured trunk-based branching with continuous deployment off of `master` was the quickest way to make sure we stay up-to-date with PRs for the checks. If there is an entire pull request etc off of a typo or updated documentation, I don't see what the reasoning is for not releasing it immediately or the downside of just incrementing a version esp when it isn't breaking.
> > However, if the majority of people want to do release branches and all that, that's fine, but in that case, I think that doing git flow with `develop` and `master` branches make more sense than merging into `master`, then branching off of master into `rc-**` and then merging `rc-**` back into `master` off of a release and potentially dealing with even more breaking changes and merge conflicts introduced into master before `rc-**` is merged back in. If we're going to use the git-flow model, continuous deployment off of master branch still makes sense.
> > Perhaps I could find a way to automatically tag `package.json` w up-to-date version using git tags?
> 
> I don't think we need release branches. I don't think merging into master should publish a new version of the library. Instead, we can have CircleCI publish to npm when we make a release. So we create 1.0.0, release that as the base version. Then we make a few changes (let's say 3), that's 3 PRs into master. We decide we want that to be 1.0.1, so we release the current master branch with release 1.0.1, which gets published to npm. Now npm has versions 1.0.0 and 1.0.1. I don't think we need to release every PR or merge into master.
> 
> This workflow is how I setup the Ruby linter: https://github.com/silvercar/silvercop/blob/master/.circleci/config.yml

Oh haha I totally forgot about the existence of tags in git. Makes sense! See latest changes? @johnny-1111one1

**Comment by johnny-1111one1** [comment]
@bucktower Yup, looks better.

**Pull request merged**